### ISSUE TITLE
Use 308 Redirect to keep Request Method

### ISSUE THREAD

**Issue by geertijewski** [certbot/certbot#7407]
letsencrypt/certbot should (ideally) "just" add HTTPS to the existing site without changing existing behaviour. This works very well for all GET requests, but as the default is an 301 Redirect. Which is not correctly implemented in the most common browsers[0]. So the idea is to change the redirect to a 308.

## My operating system is (include version):
debian buster

## I installed Certbot with (certbot-auto, OS package manager, pip, etc):
pip

## Here is the relevant nginx server block or Apache virtualhost for the domain I am configuring:
```
server {
    if ($host = example.org) {                                      
        return 301 https://$host$request_uri;
    } # managed by Certbot
...
```
## Here is the relevant nginx server block or Apache virtualhost after the change:
```
server {
    if ($host = example.org) {                                      
        return 308 https://$host$request_uri;
    } # managed by Certbot
...
```

[0] https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/301
> Even if the specification requires the method (and the body) not to be altered when the redirection is performed, not all user-agents align with it - you can still find this type of bugged software out there. It is therefore recommended to use the 301 code only as a response for GET or HEAD methods and to use the 308 Permanent Redirect for POST methods instead, as the method change is explicitly prohibited with this status.

**Comment by joohoi** [comment]
I dug this a bit, and realized that `308` code is not supported by IE that ships with Windows 7 and 8.1 (in addition to some more exotic and old browser versions). While Windows 7 is scheduled for EOL on 14th of this month, I expect that we'll see it in the wild for years to come, and Windows 8.1 will have extended support until 2023.

While `308` redirects are technically a better solution, we should prioritize the security for maximum amount of users of the sites we configure, and unfortunately it looks like using `308` instead of what we currently have won't be able to accomplish this in the foreseeable future.

**Comment by geertijewski** [comment]
> I dug this a bit, and realized that `308` code is not supported by IE that ships with Windows 7 and 8.1

Ah that is a problem, I absolutely understand the concern. How should we proceed, can we just let the issue and PR open and then look at it again in a couple years?

**Comment by joohoi** [comment]
> How should we proceed, can we just let the issue and PR open and then look at it again in a couple years?

Yeah, I think that would be the preferred way forward. I'll add these to "Wishlist" milestone, which is a catch-all one for the future.

**Comment by bmw** [comment]
I personally like to keep our list of open PRs as short as possible. Since we cannot land #7665 yet, I moved it to the [nginx-308 branch on this repo](https://github.com/certbot/certbot/tree/nginx-308) so that work is available in the future if we want it.

**Comment by kirsv-ldnet** [comment]
Could this be something that could be set with configuration parameters? We definitely would like to have option for using 308 instead of 301.

**Comment by benitogf** [comment]
would be nice to get this either changed to 308 or as an option 👍 (small bump)

**Comment by rich-kav-moa** [comment]
Also bumping the desire to have a 308 instead of a 301.

WIndows 7 is long dead now.

**Comment by maybe-hello-world** [comment]
Also worth mentioning is that Internet Explorer is retired and no longer supported since July 15, 2022. Since June Windows forcefully opens Edge instead of IE (unless turned off by GPO explicitly). 
Windows 8.1 extended support is January 10, 2023.

**Comment by samarets** [comment]
I support it

**Comment by osirisinferi** [comment]
Reopened due to renewed interest in #9865. And it's also 3 years later, so perhaps the compatibility issues mentioned in #7665 are not an issue any more? Let us discuss again :stuck_out_tongue:

**Comment by ismxilxrif** [comment]
i think at this point of time, we should reconsider this implementation, one of our deployments failed after upgrading to https bcs the post request and the body was changed to GET and the body was not following the redirect. it may have been our fault. however, i believe not many people are aware of the 301 vs 308, and 308's benefits compared to 301

hope that this suggestion can be implemented, not just from nginx's side but also apache too.

**Comment by marbens-arch** [comment]
I propose a compromise: use 301 if the request method is GET, and 308 otherwise, if compatibility is still a concern.
